### PR TITLE
add System.PosixCompat.Process

### PR DESCRIPTION
**Pull request by ners**
This PR adds the `getProcessID` function which is supported on both Unix and Win32 systems. It is already used by `lsp-test` and `lsp-client`.

### CODE REVIEW

File: src/System/PosixCompat/Process.hs [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+This module makes the operations exported by @System.Posix.Process@
+available on all platforms. On POSIX systems it re-exports operations from
+@System.Posix.Process@, on other platforms it emulates the operations as far
+as possible.
+-}
+module System.PosixCompat.Process (
+      getProcessID
+    ) where
+
+#ifndef mingw32_HOST_OS
```
Comment (andreasabel) [inline src/System/PosixCompat/Process.hs]: I'd prefer `#ifdef` over `#ifndef`.

File: src/System/PosixCompat/Process.hs [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+import System.Posix.Process
+
+#else
+
+import System.Posix.Types (ProcessID)
+
+getProcessID :: IO ProcessID
+getProcessID = fromIntegral <$> System.Win32.Process.getCurrentProcessID
```
Comment (andreasabel) [inline src/System/PosixCompat/Process.hs]: CI fails here on Windows: https://github.com/haskell-pkg-janitors/unix-compat/actions/runs/6729599930/job/18346250074?pr=6#step:5:122
``` 
src\System\PosixCompat\Process.hs:22:33: error:
Error:     Not in scope: ‘System.Win32.Process.getCurrentProcessID’
    No module named ‘System.Win32.Process’ is imported.
   |
22 | getProcessID = fromIntegral <$> System.Win32.Process.getCurrentProcessID
```

**Review by andreasabel — CHANGES_REQUESTED** [review]
Thanks for this contribution.

Please add a testcase demonstrating the use of this new module.

File: CHANGELOG.md [hunk 3]
```diff
@@ -1,6 +1,10 @@
+## Version 0.8 (2023-11-01)
```
Comment (andreasabel) [inline CHANGELOG.md]: Bump to 0.7.1 sufficient according to https://pvp.haskell.org/ paragraphs 2 and 8.

File: src/System/PosixCompat/Process.hs (hunk 1 above)
Comment (ners) [reply src/System/PosixCompat/Process.hs]: I've inverted it. I followed the example from `System.PosixCompat.Time`.

File: src/System/PosixCompat/Process.hs (hunk 2 above)
Comment (ners) [reply src/System/PosixCompat/Process.hs]: Whoops! Hopefully this fixes it.

**Comment by ners** [comment]
Thank you for the speedy review, @andreasabel. I've added a very simple test, though I'm not entirely sure what I could assert in it.
Could you please run the CI again?

File: src/System/PosixCompat/Process.hs [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+as possible.
+-}
+module System.PosixCompat.Process (
+      getProcessID
+    ) where
+
+#ifdef mingw32_HOST_OS
+
+import System.Posix.Types (ProcessID)
+import System.Win32.Process (getCurrentProcessID)
```
Comment (andreasabel) [inline src/System/PosixCompat/Process.hs]: I think this must be 'Id' not 'ID'.

File: CHANGELOG.md [hunk 5]
```diff
@@ -1,6 +1,10 @@
+## Version 0.7.1 (2023-11-01)
+
+- Add `System.PosixCompat.Process` module
```
Comment (andreasabel) [inline CHANGELOG.md]: Maybe add here:
> exporting `getProcessID`

**Review by andreasabel — COMMENTED** [review]
Thanks for adding the test!

Comment: You seem to never run this on Windows locally, so there is still a build failure due to a typo...

Anyway, I think this can be merged once CI passes.

**Comment by ners** [comment]
@andreasabel apologies for the back-and-forth, I was hoping the CI would run automatically on PR. I've now discovered that I can enable CI actions on my own fork, so in the future I can test locally before opening a PR. :slightly_smiling_face:

**Comment by ners** [comment]
@andreasabel anything else to be done here or can we merge?

File: src/System/PosixCompat/Process.hs [hunk 6]
```diff
@@ -0,0 +1,25 @@
+{-# LANGUAGE CPP #-}
+
+{-|
+This module makes the operations exported by @System.Posix.Process@
+available on all platforms. On POSIX systems it re-exports operations from
```
Comment (andreasabel) [inline src/System/PosixCompat/Process.hs]: I suppose this is a more a statement of intent than of actual function.
Currently, it only re-exports a single operation from `System.Posix.Process`.
Maybe it  could be reformulated as `This module intends to make the operations of @System.Posix.Process@ available on all platforms.`  What do you think?

**Review by andreasabel — APPROVED** [review]

**Comment by andreasabel** [comment]
Do you want this to be released asap or only eventually?

File: src/System/PosixCompat/Process.hs (hunk 6 above)
Comment (ners) [reply src/System/PosixCompat/Process.hs]: That sounds good; I’m in favour of iteratively improving the docs. Before the next version we could consider rewording the other modules as well.

**Comment by ners** [comment]
Ideally I’d like version 0.7.1 sooner rather than later so I can update lsp-client.

What I’d like to discuss afterwards is potentially refactoring the partially supported functions (e.g. the ones that throw exceptions on Windows) to a separate module in preparation for 0.8.

**Comment by andreasabel** [comment]
I removed the Haskell warning for shadowing `id`.
I could not push to your branch (maybe because you used `main` rather than a dedicated one on your fork), so I moved this PR to
- #8

**Comment by andreasabel** [comment]
Released in 0.7.1.

**Pull request closed without merging**